# Coursework 1 - Overall ML Pipeline (Stroke Prediction)

**Dataset:** healthcare-dataset-stroke-data.csv (Kaggle / UCU format)

This notebook presents a structure machine learning workflow for stroke predicition using logistic regression.
The focus is on rigorous experiment design, including dataset exploration, preprocessing, model training and validation, and performance evaluation.
Attention is given to reproducibility, imterpretability and computational efficiency with visualisations supporting the analysis and clear justification of modelling choices.


# Set up and Imports

In [18]:
# Python core and scientific libraries
import time 
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt

# Model selection and evaluation tools
from sklearn.model_selection import train_test_split, KFold, StratifiedKFold, cross_validate, learning_curve

# Pipeline and preprocessing
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, PolynomialFeatures

# Models
from sklearn.linear_model import Ridge, LogisticRegression

# Metrics
from sklearn.metrics import (
    mean_squared_error, mean_absolute_error, r2_score,
    accuracy_score, f1_score, roc_auc_score,
    confusion_matrix, ConfusionMatrixDisplay
)

# Reproducibility
RNG = 42
np.random.seed(RNG)

# Load Dataset

In [19]:
DATA_PATH = r"healthcare-dataset-stroke-data.csv" # Set the path to the stroke dataset

df = pd.read_csv(DATA_PATH, sep = ',') # Read the csv file into a DataFrame

assert "stroke" in df.columns, "Expected 'stroke' column." # Check if the DataFrame has a column named 'sroke'

print(df.shape)
df.head() # Display first 5 rows of the DataFrame to give a quick preview of the data

(5110, 12)


,id,gender,age,hypertension,heart_disease,ever_married,work_type,Residence_type,avg_glucose_level,bmi,smoking_status,stroke
0,9046,Male,67.0,0,1,Yes,Private,Urban,228.69,36.6,formerly smoked,1
1,51676,Female,61.0,0,0,Yes,Self-employed,Rural,202.21,NaN,never smoked,1
2,31112,Male,80.0,0,1,Yes,Private,Rural,105.92,32.5,never smoked,1
3,60182,Female,49.0,0,0,Yes,Private,Urban,171.23,34.4,smokes,1
4,1665,Female,79.0,1,0,Yes,Self-employed,Rural,174.12,24.0,never smoked,1


# Initial Exploration

In [20]:

display(df.describe(include='all')) # Give summary statistics for the dataset and ensures it includes both numeric and categorical columns

print("\nMissing values per column:") 

print(df.isna().sum().sort_values(ascending=False)) # See which columns might need cleaning or imputation.

,id,gender,age,hypertension,heart_disease,ever_married,work_type,Residence_type,avg_glucose_level,bmi,smoking_status,stroke
count,5110.000000,5110,5110.000000,5110.000000,5110.000000,5110,5110,5110,5110.000000,4909.000000,5110,5110.000000
unique,NaN,3,NaN,NaN,NaN,2,5,2,NaN,NaN,4,NaN
top,NaN,Female,NaN,NaN,NaN,Yes,Private,Urban,NaN,NaN,never smoked,NaN
freq,NaN,2994,NaN,NaN,NaN,3353,2925,2596,NaN,NaN,1892,NaN
mean,36517.829354,NaN,43.226614,0.097456,0.054012,NaN,NaN,NaN,106.147677,28.893237,NaN,0.048728
std,21161.721625,NaN,22.612647,0.296607,0.226063,NaN,NaN,NaN,45.283560,7.854067,NaN,0.215320
min,67.000000,NaN,0.080000,0.000000,0.000000,NaN,NaN,NaN,55.120000,10.300000,NaN,0.000000
25%,17741.250000,NaN,25.000000,0.000000,0.000000,NaN,NaN,NaN,77.245000,23.500000,NaN,0.000000
50%,36932.000000,NaN,45.000000,0.000000,0.000000,NaN,NaN,NaN,91.885000,28.100000,NaN,0.000000
75%,54682.000000,NaN,61.000000,0.000000,0.000000,NaN,NaN,NaN,114.090000,33.100000,NaN,0.000000



Missing values per column:
bmi                  201
id                     0
gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
smoking_status         0
stroke                 0
dtype: int64


# Preprocessing

# Define Features and Target

# Train-Test Split

# Model Pipeline

# Cross-Validation

# Fit Model on Training Data

# Evaluate on Test Data

# Confusion Matrix

# Reproducibility and Reporting



In [21]:
import sklearn, sys
print("Python:", sys.version)
print("NumPy:", np.__version__)
print("Pandas:", pd.__version__)
print("scikit-learn:", sklearn.__version__)

print("\nPipelines used:")
print("Regression:", reg_pipe)
print("Classification:", clf_pipe)


Python: 3.13.7 | packaged by Anaconda, Inc. | (main, Sep  9 2025, 19:54:37) [MSC v.1929 64 bit (AMD64)]
NumPy: 2.3.3
Pandas: 2.3.2
scikit-learn: 1.7.2

Pipelines used:


NameError: name 'reg_pipe' is not defined

### Mini Model Card

- **Objective:**  Predict whether a patient will experience a stroke based on health and demographic features.

- **Data:** Stroke dataset with 5110 entries; features include age, gender, hypertension, heart disease, marital status, work type, residence type, average glucose level, BMI, and smoking status; CSV is comma-separated.

- **Validation:** 5-fold stratified cross-validation on the training set; stratified train-test split to preserve class balance; final evaluation on an untouched test set.

- **Key metrics:** Reported metrics include accuracy, F1 score, and ROC-AUC. Example numbers 

- **Strengths / limitations:** Logistic regression provides interpretability; however, class imbalance (~5% stroke cases) may reduce sensitivity for minority class. Preprocessing required for missing BMI values and categorical encoding.

- **Ethical notes / misuse risks:** Predictions depend on dataset demographics and may not generalize to other populations; model is not a medical diagnostic tool and should not be used for clinical decision-making without expert oversight.